### ISSUE TITLE
A bug for concave features

### ISSUE THREAD

**Issue by WANGYINGYU** [JiahengZhao/2Dlidar_simulation_fs#1]
Hi Jiaheng,

Thank you very much for your simulator code which help me a lot for my paper. However, I found a important bug in the code which occurs on all big concave features. For the concave areas of these features, the laser will strike through the outermost wall to the inner wall. The bug is caused by the code only considering at most 2 hit points, when in fact there are not at most 2 hit points for concave shape features. Only a small change to the function intersectionFS() can solve this problem. Below is the code how I solved this bug.

```
function [fval, info] = intersectionFS(F,an,bn,ang,x0,y0)

center=F;
N=length(an)-1;
e = 1e-10;
% tangent

theta_final=[];    
nCut = 36;
stepA = 2*pi/nCut;
xL=-pi; xU=-pi+stepA;
if abs(ang-pi/2) <= 1e-5 || abs(ang+pi/2)<= 1e-5
    for ts = 1:nCut
        fL=func_s(an,bn,xL+(ts-1)*stepA,center,x0);
        fU=func_s(an,bn,xU+(ts-1)*stepA,center,x0);
        if fL*fU>0
            Nocross=0;
        else
            theta1=dichotomyRoot_s(xL+(ts-1)*stepA,xU+(ts-1)*stepA,e,an,bn,center,x0);
            theta_final=[theta_final;theta1];
        end
    end
  
    if isempty(theta_final)
        fval=[];
        info=0;
    else
        nn = size(theta_final,1);
        D = zeros(nn,1);
%         D = zeros(2,1);
        for i = 1:N+1
%         for i = 1:N
            D = D + an(i).*cos((i-1).*theta_final) + bn(i).*sin((i-1).*theta_final);
        end
        XX=D.*cos(theta_final)+center(1);
        YY=D.*sin(theta_final)+center(2);

        all_dist = [];
        for ddd=1:nn
            temp_dist = (YY(ddd)-y0)^2+(XX(ddd)-x0)^2;
            all_dist = [all_dist,temp_dist];
        end

        [min_dist,min_id] = min(all_dist);
        fval=[XX(min_id);YY(min_id)];

        
%         dist1 = (YY(1)-y0)^2+(XX(1)-x0)^2;
%         dist2 = (YY(2)-y0)^2+(XX(2)-x0)^2;
%         if dist1<=dist2
%             fval=[XX(1);YY(1)];
%         else
%             fval=[XX(2);YY(2)];
%         end
        info=1;
    end

else % general case
    k = tan(ang);
    for ts = 1:nCut
        fL=func(an,bn,xL+(ts-1)*stepA,center,k,x0,y0);
        fU=func(an,bn,xU+(ts-1)*stepA,center,k,x0,y0);
        if fL*fU>0
            Nocross=0;
        else
            theta1=dichotomyRoot(xL+(ts-1)*stepA,xU+(ts-1)*stepA,e,an,bn,center,k,x0,y0);
            theta_final=[theta_final;theta1];
        end
    end
    
    if isempty(theta_final)
        fval=[];
        info=0;
    else
        nn = length(theta_final);
        D = zeros(length(theta_final),1);
        bn = [0;bn];
%         for i = 1:N+1
        for i = 1:N+1
            D = D + an(i).*cos((i-1).*theta_final) + bn(i).*sin((i-1).*theta_final);
        end
        XX=D.*cos(theta_final)+center(1);
        YY=D.*sin(theta_final)+center(2);
        all_ang = [];
        all_dist = [];
        for ddd=1:nn
            temp_ang = atan2(YY(ddd)-y0,XX(ddd)-x0);
            all_ang = [all_ang,temp_ang];
            temp_dist = (YY(ddd)-y0)^2+(XX(ddd)-x0)^2;
            all_dist = [all_dist,temp_dist];
        end
        diff_temp = abs(all_ang - ang);
        min_diff_temp = min(diff_temp);
        
%         temp1 =  atan2(YY(2)-y0,XX(2)-x0);
        if min_diff_temp < 1e-5
%         if abs(temp1 - ang) < 1e-5
            info = 1;
            kkkkk=1;
        else
            info = 0;
            fval = [];
            return
        end
        [min_dist,min_id] = min(all_dist);
        fval=[XX(min_id);YY(min_id)];
        
%         dist1 = (YY(1)-y0)^2+(XX(1)-x0)^2;
%         dist2 = (YY(2)-y0)^2+(XX(2)-x0)^2;
%         if dist1<=dist2
%             fval=[XX(1);YY(1)];
%         else
%             fval=[XX(2);YY(2)];
%         end


    end
end
end
```

Hope it helps with this program.

Regards,
Yingyu

**Comment by JiahengZhao** [comment]
Thanks @WANGYINGYU !  I knew the assumption of two intersections but don't have chances to update it : ) 

I will close this issue with the commit [05a7fd5](https://github.com/JiahengZhao/2Dlidar_simulation_fs/commit/05a7fd531cb187c63c2b5df9d9033ce2fa20463b).

Please pull the latest commit and let me know if the updates work for you.

**Comment by WANGYINGYU** [comment]
> Thanks @WANGYINGYU ! I knew the assumption of two intersections but don't have chances to update it : )
> 
> I will close this issue with the commit [05a7fd5](https://github.com/JiahengZhao/2Dlidar_simulation_fs/commit/05a7fd531cb187c63c2b5df9d9033ce2fa20463b).
> 
> Please pull the latest commit and let me know if the updates work for you.

The new code is better, thanks. Some other possible problem are as follows.

1. The line 9. nCut = 12 may not be accurate enough for very large irregular shapes, I recommend increasing this value and I set the value as 36. 
2. The line 12. if abs(ang-pi/2) <= 1e-5 || abs(ang+pi/2)<= 1e-5. The new code may cause an unexpected error which lead the number of an and bn to be different. I found that the original method of judging i.e. strictly equal to -pi/2 and pi/2 is more appropriate.

Hope this helps with the code.
